### PR TITLE
Add MTL logger tools

### PR DESCRIPTION
**Pull request by bhancock8**
Port over from MeTaL/mmtl and Emmental

- LogManager checks when it's time to log, evaluate, and/or checkpoint
- LogWriter/TensorboardWriter write logs
- Checkpointer handles bookeeping, saving, and loading of best model

**Test plan**
- Ported existing tests for LogManager
- Checkpointer tests will be added once a model is defined that can be saved/loaded

### CODE REVIEW

File: snorkel/mtl/loggers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+logging_default_config = {
```
Comment (henryre) [inline snorkel/mtl/loggers/__init__.py]: Let's get this outside of init

File: snorkel/mtl/loggers/log_manager.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
+import logging
+import os
+from datetime import datetime
+from typing import Dict
+
+import torch
+
+from snorkel.mtl.loggers import logging_default_config
```
Comment (henryre) [inline snorkel/mtl/loggers/log_manager.py]: e.g. `from .checkpointer import ...` is fine here

File: snorkel/types/logger.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from typing import Any, Dict
+
+# A config dict is a nested dict with a max depth of 3
+Config = Dict[str, Any]
+Path = str
```
Comment (henryre) [inline snorkel/types/logger.py]: this might be confusing to people since `Path` objects exist elsewhere. can we just use str in-place?

**Comment by henryre** [comment]
Also needs to be rebased onto redux

File: snorkel/mtl/loggers/log_manager.py (hunk 2 above)
Comment (bhancock8) [reply snorkel/mtl/loggers/log_manager.py]: PEP8 recommends absolute imports ("Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured"). Will change here for now though to be consistent with the files you've already moved over. Can discuss and update later if need be.

File: snorkel/types/logger.py (hunk 3 above)
Comment (bhancock8) [reply snorkel/types/logger.py]: Yep, will remove the `Path` type for now, though in a future commit I suggest we call-out where strings are paths explicitly.

**Pull request merged**